### PR TITLE
DSA cdf tests

### PR DESCRIPTION
**Pull request by karel-m**
Another crypto test bench https://github.com/kudelskisecurity/cdf

Good news - our RSA implementation seems to pass. But we have couple of DSA related troubles.

The first commit fixes failures when ltc is on "validating" side. There are still some failures when ltc is on "signing" side, I'll try to analyze those later.

### CODE REVIEW

**Comment by karel-m** [comment]
Well, it turned out that the case when ltc is on "signing" side is OK. So this PR is complete.

**Comment by karel-m** [comment]
@sjaeckel ping

**Comment by sjaeckel** [comment]
We could e.g. have an internal (say `LTC_SOURCE` protected) `dsa_check_key()` which is then called from within `dsa_verify_key()`, `dsa_set_pqg_dsaparam()` and `dsa_set_pqg()` (and probably others that I missed?).

**Comment by karel-m** [comment]
something like this?

**Comment by sjaeckel** [comment]
Do we really need such extensive tests? (the 2 exptmod's?)
Besides that I realized that the name `verify_key()` is nonsense, therefore I've pushed my proposal to the pr/fix-dsa-cdf-2 branch. What do you think?

1. Can we make `dsa_int_validate_key()` a bit less computational expensive by moving the 2 exptmods to `dsa_validate_key()`?
2. Wouldn't it make sense to use `LTC_MILLER_RABIN_REPS` for the prime checks?
3. With your last commit you removed stuff that you added in the commit before (in `dsa_set_pqg()`)
a. Was that accidental or itentional?
b. The 'newly added' was three `mp_cmp_d()` whereas `dsa_verify_key_ex()` has only two + a different logic, can you please check back which is correct!?

**Comment by karel-m** [comment]
we should perhaps keep the name `dsa_verify_key` as it is part of 1.17 API

**Comment by sjaeckel** [comment]
> we should perhaps keep the name `dsa_verify_key` as it is part of 1.17 API

damn, sorry - sure

I'll just update the branch

**Comment by karel-m** [comment]
> With your last commit you removed stuff that you added in the commit before (in `dsa_set_pqg()`)

The thing is that at the end of `dsa_set_pqg()` we do not have a complete key, therefore we cannot do the key check/validation.

**Comment by sjaeckel** [comment]
> The thing is that at the end of `dsa_set_pqg()` we do not have a complete key, therefore we cannot do the key check/validation.

but it looks to me like part of it was required for the cdf tests to pass, right?

**Comment by sjaeckel** [comment]
> I'll just update the branch

pushed

**Comment by karel-m** [comment]
What we need is to avoid loading/setting/creating a DSA key with `p` or `g` or `q` (not sure what was the main trouble maker) with value `0`. 

It would be better to check it in `dsa_set_pqg` but it is IMO fine to do it a bit later during `dsa_set_key` (where we have the key complete).

**Comment by karel-m** [comment]
I have slightly updated your pr/fix-dsa-cdf-2 - which works and can be merged

**Comment by sjaeckel** [comment]
> What we need is to avoid loading/setting/creating a DSA key with p or g or q (not sure what was the main trouble maker) with value 0.

Okay, but the different parts of the implementation I saw had checks for `{p,g,q} > 1` `|p| > 1` and I don't know what else. Reading quickly through FIPS 186-4 I couldn't determine what is the correct approach so can you please check again what we really have to do?

I think it'd make sense to have 2 private functions `dsa_int_validate_group_params()` and `dsa_int_validate_key()` and 1 API function `dsa_verify_key()`. This allows us to correctly verify each step of the key import/generation.

Flow could be something like this:

```
int dsa_verify_key() {
 check_prime(q);
 check_prime(p);
 return dsa_int_validate_key();
}

int dsa_int_validate_group_params() {
 ...
}

int dsa_int_validate_key() {
 if (dsa_int_validate_group_params()) return error;
 ...
}
```

What do you think?

**Comment by karel-m** [comment]
I'll check FIPS 186-4 but meanwhile another approach for `dsa_int_validate_*`

**Comment by karel-m** [comment]
@sjaeckel could you please review this PR again?

File: src/pk/dsa/dsa_verify_key.c [hunk 1]
```diff
@@ -16,81 +16,183 @@
...
+   @param key   The key to validate
+   @param stat  [out]  Result of test, 1==valid, 0==invalid
+   @return CRYPT_OK if successful
+*/
+int dsa_int_validate_pqg(dsa_key *key, int *stat)
+{
+   void *tmp1, *tmp2;
+   int  err;
+
+   *stat = 0;
```
Comment (sjaeckel) [inline src/pk/dsa/dsa_verify_key.c]: ew, de-referencing stat before checking for NULL

File: src/headers/tomcrypt_pk.h [hunk 2]
```diff
@@ -479,7 +479,11 @@ int dsa_decrypt_key(const unsigned char *in,  unsigned long  inlen,
 int dsa_import(const unsigned char *in, unsigned long inlen, dsa_key *key);
 int dsa_export(unsigned char *out, unsigned long *outlen, int type, dsa_key *key);
 int dsa_verify_key(dsa_key *key, int *stat);
-
+#ifdef LTC_SOURCE
```
Comment (sjaeckel) [inline src/headers/tomcrypt_pk.h]: can you please add the obligatory `/* internal helper functions */` comment

File: src/pk/dsa/dsa_verify_key.c [hunk 3]
```diff
@@ -16,81 +16,183 @@
...
+   if ((err = mp_exptmod(key->g, key->q, key->p, tmp1)) != CRYPT_OK) { goto error; }
+   if (mp_cmp_d(tmp1, 1) != LTC_MP_EQ) {
+      err = CRYPT_OK;
+      goto error;
+   }
+
+   err   = CRYPT_OK;
+   *stat = 1;
+error:
+   mp_clear_multi(tmp1, tmp2, NULL);
```
Comment (sjaeckel) [inline src/pk/dsa/dsa_verify_key.c]: can you please (always) reverse the order when clearing MPI's

File: src/pk/dsa/dsa_import.c [hunk 4]
```diff
@@ -116,10 +116,21 @@ int dsa_import(const unsigned char *in, unsigned long inlen, dsa_key *key)
...
+   key->qord = mp_unsigned_bin_size(key->q);
 
-  if (key->qord >= LTC_MDSA_MAX_GROUP || key->qord <= 15 ||
-      (unsigned long)key->qord >= mp_unsigned_bin_size(key->p) || (mp_unsigned_bin_size(key->p) - key->qord) >= LTC_MDSA_DELTA) {
+   /* quick p, q, g validation, without primality testing */
+   if ((err = dsa_int_validate_pqg(key, &stat)) != CRYPT_OK) {
+      goto LBL_ERR;
+   }
+   if (stat == 0) {
+      err = CRYPT_INVALID_PACKET;
```
Comment (sjaeckel) [inline src/pk/dsa/dsa_import.c]: Looks like we inconsistently return either `CRYPT_INVALID_PACKET` or `CRYPT_INVALID_ARG`, we should always use the same error code. I'd say we should go for `CRYPT_INVALID_PACKET` (an alternative could be to introduce a new error code, but not sure if that's really necessary).

File: src/pk/dsa/dsa_verify_key.c [hunk 5]
```diff
@@ -16,81 +16,183 @@
...
+   int err, res;
+
    *stat = 0;
+   LTC_ARGCHK(key  != NULL);
+   LTC_ARGCHK(stat != NULL);
 
-   /* first make sure key->q and key->p are prime */
-   if ((err = mp_prime_is_prime(key->q, 8, &res)) != CRYPT_OK) {
+   /* key->q prime? */
+   if ((err = mp_prime_is_prime(key->q, LTC_MILLER_RABIN_REPS, &res)) != CRYPT_OK) {
```
Comment (sjaeckel) [inline src/pk/dsa/dsa_verify_key.c]: :+1:

**Comment by karel-m** [comment]
Oh sorry, I should perhaps wait for Travis-CI

**Comment by sjaeckel** [comment]
Too late, the build's going to fail ;)

**Comment by sjaeckel** [comment]
but nvm, as the merge will be built as well...

**Pull request merged**